fastcrypto-zkp/src/bn254/zk_login.rs

```diff
@@ -0,0 +1,749 @@
+    let base64_url_character_set =
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
+    input
+        .chars()
+        .flat_map(|c| {
+            let index = base64_url_character_set.find(c).unwrap();
+            let mut bits = Vec::new();
+            for i in 0..6 {
+                bits.push((index >> (5 - i)) & 1 == 1);
+            }
+            bits
+        })
+        .collect()
+}
+
+/// Convert a bitarray to a bytearray.
+fn bits_to_bytes(bits: &[bool]) -> Vec<u8> {
+    let mut bytes: Vec<u8> = Vec::new();
+    let mut current_byte: u8 = 0;
+    let mut bits_remaining: u8 = 8;
+
+    for bit in bits.iter() {
+        if *bit {
+            current_byte |= 1 << (bits_remaining - 1);
+        }
+        bits_remaining -= 1;
+        if bits_remaining == 0 {
+            bytes.push(current_byte);
+            current_byte = 0;
+            bits_remaining = 8;
+        }
+    }
+
+    if bits_remaining < 8 {
+        bytes.push(current_byte);
+    }
+
+    bytes
+}
+
+/// Convert a big int string to a big endian bytearray.
+pub fn big_int_str_to_bytes(value: &str) -> Vec<u8> {
+    BigInt::from_str(value)
+        .expect("Invalid big int string")
+        .to_bytes_be()
+        .1
+}
+
+/// Calculate the integer value from the bytearray.
+fn calculate_value_from_bytearray(arr: &[u8]) -> usize {
+    let sized: [u8; 8] = arr.try_into().expect("Invalid byte array");
+    ((sized[7] as u16) | (sized[6] as u16) << 8).into()
+}
+
+/// Given a chunk of bytearray, parse it as an ascii string and decode as a JWTHeader.
+/// Return the JWTHeader if its fields are valid.
+fn parse_and_validate_header(chunk: &[u8]) -> Result<JWTHeader, FastCryptoError> {
+    let header_str = std::str::from_utf8(chunk)
+        .map_err(|_| FastCryptoError::GeneralError("Cannot parse header string".to_string()))?;
+    let decoded_header = Base64UrlUnpadded::decode_vec(header_str)
+        .map_err(|_| FastCryptoError::GeneralError("Invalid jwt header".to_string()))?;
+    let json_header: Value = serde_json::from_slice(&decoded_header)
+        .map_err(|_| FastCryptoError::GeneralError("Invalid json".to_string()))?;
+    let header: JWTHeader = serde_json::from_value(json_header)
+        .map_err(|_| FastCryptoError::GeneralError("Cannot parse jwt header".to_string()))?;
+    if header.alg != "RS256" || header.typ != "JWT" {
+        Err(FastCryptoError::GeneralError("Invalid header".to_string()))
+    } else {
+        Ok(header)
+    }
+}
+
+/// Calculate the merklized hash of the given bytes after 0 paddings.
+pub fn calculate_merklized_hash(bytes: &[u8]) -> String {
+    let mut bitarray = bytearray_to_bits(bytes);
+    pad_bitarray(&mut bitarray, 248);
+    let bigints = convert_to_bigints(&bitarray, 248);
+    let mut poseidon1 = PoseidonWrapper::new(15);
+    let hash1 = poseidon1.hash(&bigints[0..15]);
+
+    let mut poseidon2 = PoseidonWrapper::new(bigints.len() - 15);
```

Comment:
What happens if bigints.len() < 15? 

Minor: You are combining two different operations (base change and poseidon hash) into one function (which is fine but perhaps warrants different function?).